### PR TITLE
Implement fargate profile

### LINKED ISSUE #407: Implement EKS Fargate Profile
### What problem are you facing?

I would like to be able to schedule pods on Fargate: https://docs.aws.amazon.com/eks/latest/userguide/fargate.html.

### How could Crossplane help solve your problem?

Implement `kind: FargateProfile, apiVersion: eks.aws.crossplane.io`

### PR DESCRIPTION
**Pull request by Knappek**
### Description of your changes

Fixes #407 

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

1. Manual testing (example/eks)
2. Unit testing for the EKS FargateProfile controller and client

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: examples/eks/fargateprofile.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+apiVersion: eks.aws.crossplane.io/v1alpha1
+kind: FargateProfile
+metadata:
+  name: my-fargateprofile
+  labels:
+    examplelabelkey: "examplelabelval"
+spec:
+  forProvider:
+    region: eu-central-1
```
Comment (muvaf) [inline examples/eks/fargateprofile.yaml:9]:
```suggestion
    region: us-east-1
```

Would you mind changing this to `us-east-1`? Most of the examples are in that region, which makes it easier to reference each other.

File: examples/eks/fargateprofile.yaml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    clusterName: my-awesome-cluster
+    podExecutionRoleRef:
+      name: fargate-pod-execution-role
```
Comment (muvaf) [inline examples/eks/fargateprofile.yaml:12]: I'd go for the name of `IAMRole` example we have. Similarly for `clusterNameRef.name`.

File: pkg/clients/eks/fargateprofile_test.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+package eks
+
+import (
+	"testing"
+	// "time"
+
+	"github.com/aws/aws-sdk-go-v2/service/eks"
+	"github.com/google/go-cmp/cmp"
+
+	// 	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (muvaf) [inline pkg/clients/eks/fargateprofile_test.go:26]:
```suggestion
```

File: pkg/clients/eks/fargateprofile_test.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package eks
+
+import (
+	"testing"
+	// "time"
```
Comment (muvaf) [inline pkg/clients/eks/fargateprofile_test.go:21]:
```suggestion
```

File: pkg/controller/eks/fargateprofile/controller.go [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+import (
+	"context"
+	"reflect"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/pkg/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	awseks "github.com/aws/aws-sdk-go-v2/service/eks"
```
Comment (muvaf) [inline pkg/controller/eks/fargateprofile/controller.go:28]:
```suggestion
	awseks "github.com/aws/aws-sdk-go-v2/service/eks"
	
```

File: pkg/controller/eks/fargateprofile/controller.go [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+	case v1alpha1.FargateProfileStatusDeleting:
+		cr.Status.SetConditions(runtimev1alpha1.Deleting())
+	default:
+		cr.Status.SetConditions(runtimev1alpha1.Unavailable())
+	}
+
+	return managed.ExternalObservation{
+		ResourceExists:          true,
+		ResourceUpToDate:        eks.IsFargateProfileUpToDate(cr.Spec.ForProvider, rsp.FargateProfile),
+		ResourceLateInitialized: !reflect.DeepEqual(current, &cr.Spec.ForProvider),
```
Comment (muvaf) [inline pkg/controller/eks/fargateprofile/controller.go:119]: How about using `cmp.Equal` here for consistency?

**Review by muvaf — COMMENTED** [review]
We made a pass over the PR in a private repo and I think it's looking pretty good. There are a few small things but otherwise it should be ready to merge. Thanks @Knappek !

**Comment by muvaf** [comment]
Oh and it'd be great if you can squeeze some of the commit history to have fewer commits.

File: apis/eks/v1alpha1/fargate_profile_types.go [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+type FargateProfileSelector struct {
+
+	// The Kubernetes labels that the selector should match. A pod must contain
+	// all of the labels that are specified in the selector for it to be considered
+	// a match.
+	Labels map[string]string `json:"labels,omitempty"`
+
+	// The Kubernetes namespace that the selector should match.
+	Namespace *string `json:"namespace,omitempty"`
+	// contains filtered or unexported fields
```
[the remark this answers is not in the document]
Comment (Knappek) [reply apis/eks/v1alpha1/fargate_profile_types.go:47]: I don't know :D. I will remove it.

File: pkg/controller/eks/fargateprofile/controller.go (hunk 6 above)
Comment (Knappek) [reply pkg/controller/eks/fargateprofile/controller.go]: done

File: pkg/controller/eks/fargateprofile/controller.go (hunk 5 above)
Comment (Knappek) [reply pkg/controller/eks/fargateprofile/controller.go]: done

File: pkg/clients/eks/fargateprofile_test.go (hunk 4 above)
Comment (Knappek) [reply pkg/clients/eks/fargateprofile_test.go]: done

File: pkg/clients/eks/fargateprofile_test.go (hunk 3 above)
Comment (Knappek) [reply pkg/clients/eks/fargateprofile_test.go]: done

File: examples/eks/fargateprofile.yaml (hunk 2 above)
Comment (Knappek) [reply examples/eks/fargateprofile.yaml]: done

File: examples/eks/fargateprofile.yaml (hunk 1 above)
Comment (Knappek) [reply examples/eks/fargateprofile.yaml]: done

File: apis/eks/v1alpha1/fargate_profile_types.go [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+	// The Amazon Resource Name (ARN) of the pod execution role to use for pods
+	// that match the selectors in the Fargate profile. The pod execution role allows
+	// Fargate infrastructure to register with your cluster as a node, and it provides
+	// read access to Amazon ECR image repositories. For more information, see Pod
+	// Execution Role (https://docs.aws.amazon.com/eks/latest/userguide/pod-execution-role.html)
+	// in the Amazon EKS User Guide.
+	//
+	// PodExecutionRoleArn is a required field
+	// +immutable
+	PodExecutionRoleArn string `json:"podExecutionRoleArn"`
```
[the remark this answers is not in the document]
Comment (Knappek) [reply apis/eks/v1alpha1/fargate_profile_types.go]: done

**Review by muvaf — APPROVED** [review]
@Knappek Thank you!

**Pull request merged**